chia/wallet/wallet_node.py

```diff
@@ -1022,25 +1008,29 @@ def get_full_node_peers_in_order(self) -> List[WSChiaConnection]:
                 neither.append(node)
         return synced_and_trusted + synced + trusted + neither
 
-    async def get_timestamp_for_height(self, height: uint32) -> uint64:
+    async def get_timestamp_for_height_from_peer(self, height: uint32, peer: WSChiaConnection) -> Optional[uint64]:
         """
         Returns the timestamp for transaction block at h=height, if not transaction block, backtracks until it finds
         a transaction block
         """
-        for cache in self.peer_caches.values():
-            cache_ts: Optional[uint64] = cache.get_height_timestamp(height)
-            if cache_ts is not None:
-                return cache_ts
+        cache = self.get_cache_for_peer(peer)
+        cached_timestamp = cache.get_height_timestamp(height)
+        if cached_timestamp is not None:
+            return cached_timestamp
 
-        for peer in self.get_full_node_peers_in_order():
-            last_tx_block = await fetch_last_tx_from_peer(height, peer)
-            if last_tx_block is None:
-                continue
+        last_tx_block = await fetch_last_tx_from_peer(height, peer)
```

Comment:
"last_tx" sounds like the most recent, but you specify a height, which suggests that you get the block at the specified height. Which one is it?

Reply:
There is some heavy walk back logic backed into `fetch_last_tx_from_peer`. See https://github.com/Chia-Network/chia-blockchain/blob/208485075c335cfe8a715fad10192b68a0afca25/chia/wallet/util/wallet_sync_utils.py#L51C11-L66

Thats what i want to get rid of with having the last timestamp included in the initial message with the new protocol and also why #15500 returns them.

Reply:
I think the function you link to also has a confusing name. what does "last_tx" refer to exactly? It doesn't return a transaction, which typically is what "tx" refers to. It returns a `HeaderBlock`, but it's not even the "last" header block, it's the header block at the specified height.

perhaps I'm misunderstanding something, but this function should be called something like; `fetch_header_at_height()`.

I suppose `TX-block` refers to transaction block, and "last" refers to the most recent TX-block before or at `height`